### PR TITLE
WIP : support for hloc toolbox (superpoint, superglue, netvlad,...)

### PR DESCRIPTION
**Pull request by RandomPrototypes**
This adds support for [hloc](https://github.com/cvg/Hierarchical-Localization) which brings support for more recent features and matches such as superpoint and superglue.

I added command-line options to select the sfm toolbox (colmap, hloc), feature (sift, superpoint,...), and matcher (NN, superglue,...).

It requires hloc as dependency.

### CODE REVIEW

File: nerfstudio/process_data/hloc_utils.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
+from pathlib import Path
+from typing_extensions import Literal
+
+from nerfstudio.process_data.process_data_utils import CameraModel
+from hloc import extract_features, match_features, reconstruction, visualization, pairs_from_retrieval, pairs_from_retrieval, pairs_from_exhaustive
+
+from hloc.triangulation import (import_features, import_matches)
+from hloc.utils.io import get_keypoints, get_matches
+
+import pycolmap
```
Comment (tancik) [inline nerfstudio/process_data/hloc_utils.py:10]: We don't have this as a dependency

File: nerfstudio/process_data/hloc_utils.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
+from pathlib import Path
+from typing_extensions import Literal
+
+from nerfstudio.process_data.process_data_utils import CameraModel
+from hloc import extract_features, match_features, reconstruction, visualization, pairs_from_retrieval, pairs_from_retrieval, pairs_from_exhaustive
+
+from hloc.triangulation import (import_features, import_matches)
```
Comment (tancik) [inline nerfstudio/process_data/hloc_utils.py:7]: These should be wrapped in a try/except and notify the user how to install if they are missing.

File: nerfstudio/process_data/hloc_utils.py (hunk 1 above)
Comment (RandomPrototypes) [reply nerfstudio/process_data/hloc_utils.py:10]: Is it possible to add these dependencies?
or should I make it as a separate script that can be called if the user installed the additional dependencies manually?

**Review by tancik — COMMENTED** [review]
Thanks for adding support for hloc. My main comment atm is around notifying users how to install hloc if it is not included. We should also add documentation for these alternative matches in the web docs.

File: nerfstudio/process_data/hloc_utils.py (hunk 2 above)
Comment (RandomPrototypes) [reply nerfstudio/process_data/hloc_utils.py:7]: ok, I'll look at it.

File: nerfstudio/process_data/hloc_utils.py (hunk 1 above)
Comment (tancik) [reply nerfstudio/process_data/hloc_utils.py:10]: hloc isn't provided as a pypi package which unfortunately means that we can't include it in our pypi package. We could add something like this - https://github.com/nerfstudio-project/nerfstudio/blob/main/nerfstudio/utils/install_checks.py

**Comment by RandomPrototypes** [comment]
ok, I'll prepare a new version with these try/except for dependency, notification, and documentation update

File: nerfstudio/process_data/hloc_utils.py (hunk 1 above)
Comment (tancik) [reply nerfstudio/process_data/hloc_utils.py:10]: Also is `pycolmap` necessary? We include many colmap utils here - https://github.com/nerfstudio-project/nerfstudio/blob/main/nerfstudio/process_data/colmap_utils.py

**Comment by RandomPrototypes** [comment]
I updated the PR :
hloc is now an optional dependency. If it is not installed, it will not cause any error except if we try to use a feature from it (superpoint, superglue,...), it will output a red error message : "Error: To use this set of parameters (superpoint_aachen/superglue/hloc), you must install hloc toolbox!!"

The documentation of ns-process-data has a new line "You may also want to install hloc for more feature detector and matcher options." with a link to the github page. The new cli arguments (sfm_tool, feature_type, matcher_type) have a description in the doc too.

File: nerfstudio/process_data/hloc_utils.py (hunk 1 above)
Comment (RandomPrototypes) [reply nerfstudio/process_data/hloc_utils.py]: pycolmap is a dependency of hloc, and the pycolmap enums are used to configure the reconstruction

File: scripts/process_data.py [hunk 3]
```diff
@@ -28,6 +29,30 @@
 CONSOLE = Console(width=120)
 
 
+def find_tool_feature_matcher_combination(sfm_tool, feature_type, matcher_type):
```
Comment (tancik) [inline scripts/process_data.py:32]: Can you add type hints.

File: scripts/process_data.py [hunk 4]
```diff
@@ -28,6 +29,30 @@
...
+    """Find a valid combination of sfm tool, feature type, and matcher type.
+    Basically, replace the default parameters 'any' by usable value"""
```
Comment (tancik) [inline scripts/process_data.py:34]: The `Args` and `Returns` should be documented.

File: scripts/process_data.py (hunk 3 above)
Comment (tancik) [inline scripts/process_data.py:32]: It might make sense to move this fn to `process_data_utils.py`

**Review by tancik — COMMENTED** [review]
Thanks for the updates! Almost ready, just a few small additional comments.

File: scripts/process_data.py (hunk 3 above)
Comment (RandomPrototypes) [reply scripts/process_data.py]: done

File: scripts/process_data.py (hunk 4 above)
Comment (RandomPrototypes) [reply scripts/process_data.py]: done

File: scripts/process_data.py (hunk 3 above)
Comment (RandomPrototypes) [reply scripts/process_data.py]: done

File: scripts/process_data.py [hunk 5]
```diff
@@ -127,7 +103,7 @@ def main(self) -> None:
         if not self.skip_colmap:
             colmap_dir.mkdir(parents=True, exist_ok=True)
 
-            (sfm_tool, feature_type, matcher_type) = find_tool_feature_matcher_combination(
+            (sfm_tool, feature_type, matcher_type) = process_data_utils.find_tool_feature_matcher_combination(
```
Comment (tancik) [inline scripts/process_data.py:106]: Can you raise an error if no valid tool is found.

File: scripts/process_data.py [hunk 6]
```diff
@@ -251,7 +227,7 @@ def main(self) -> None:
...
```
Comment (tancik) [inline scripts/process_data.py:230]: Same here

File: scripts/process_data.py (hunk 5 above)
Comment (RandomPrototypes) [reply scripts/process_data.py:106]: find_tool_feature_matcher_combination returns (None,None,None) when no valid tool is found, which brings to the message CONSOLE.log("[bold yellow]Warning: Invalid combination of sfm_tool, feature_type, and matcher_type ").

I'll add a sys.exit(1) after the message is printed.

Comment (RandomPrototypes) [reply scripts/process_data.py:106]: I changed the message to "[bold red]Invalid combination of sfm_tool, feature_type, and matcher_type, exiting" and added sys.exit(1)

File: scripts/process_data.py (hunk 6 above)
Comment (RandomPrototypes) [reply scripts/process_data.py:231]: same as other message

**Review by tancik — APPROVED** [review]
LGTM, Thanks for adding this functionality! Sorry for the delayed reviews.

**Comment by Diana-Zeng** [comment]
When I use the command:
ns-process-data images --data /media/diana/001/c \
                                                  --output /home/diana/NeRF/nerfstudio/selfdata/20230404/cc \
                                                  --camera perspective \
                                                  --matching vocab_tree \
                                                  --sfm-tool hloc \
                                                  --feature-type superpoint_aachen \
                                                  --matcher-type superglue \
                                                  --colmap-cmd /home/diana/COLMAP/colmap/build/src/exe/colmap

This error occurred:
**scipy.io.matlab._miobase.MatReadError: Mat file appears to be empty.**

These are complete information:
process_data_utils.py:363
[2023/04/10 18:58:47 hloc INFO] Extracting local features with configuration:
{'model': {'max_keypoints': 4096, 'name': 'superpoint', 'nms_radius': 3},
'output': 'feats-superpoint-n4096-r1024',
'preprocessing': {'grayscale': True, 'resize_max': 1024}}
[2023/04/10 18:58:47 hloc INFO] Skipping the extraction.
[2023/04/10 18:58:47 hloc INFO] Extracting local features with configuration:
{'model': {'name': 'netvlad'},
'output': 'global-feats-netvlad',
'preprocessing': {'resize_max': 1024}}
[2023/04/10 18:58:47 hloc INFO] Found 1440 images in root /home/diana/NeRF/nerfstudio/selfdata/20230404/cc/images.
Traceback (most recent call last):
File "/home/diana/anaconda3/envs/nerfstudio/bin/ns-process-data", line 8, in
sys.exit(entrypoint())
File "/home/diana/NeRF/nerfstudio/scripts/process_data.py", line 833, in entrypoint
tyro.cli(Commands).main()
File "/home/diana/NeRF/nerfstudio/scripts/process_data.py", line 165, in main
self._run_colmap(image_dir, colmap_dir)
File "/home/diana/NeRF/nerfstudio/scripts/process_data.py", line 231, in _run_colmap
hloc_utils.run_hloc(
File "/home/diana/NeRF/nerfstudio/nerfstudio/process_data/hloc_utils.py", line 110, in run_hloc
retrieval_path = extract_features.main(retrieval_conf, image_dir, outputs)
File "/home/diana/anaconda3/envs/nerfstudio/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
return func(*args, **kwargs)
File "/home/diana/Tools/Hierarchical-Localization/hloc/extract_features.py", line 251, in main
model = Model(conf['model']).eval().to(device)
File "/home/diana/Tools/Hierarchical-Localization/hloc/utils/base_model.py", line 17, in init
self._init(conf)
File "/home/diana/Tools/Hierarchical-Localization/hloc/extractors/netvlad.py", line 82, in _init
mat = loadmat(checkpoint, struct_as_record=False, squeeze_me=True)
File "/home/diana/anaconda3/envs/nerfstudio/lib/python3.8/site-packages/scipy/io/matlab/_mio.py", line 225, in loadmat
MR, _ = mat_reader_factory(f, **kwargs)
File "/home/diana/anaconda3/envs/nerfstudio/lib/python3.8/site-packages/scipy/io/matlab/_mio.py", line 74, in mat_reader_factory
mjv, mnv = _get_matfile_version(byte_stream)
File "/home/diana/anaconda3/envs/nerfstudio/lib/python3.8/site-packages/scipy/io/matlab/_miobase.py", line 234, in _get_matfile_version
raise MatReadError("Mat file appears to be empty")
scipy.io.matlab._miobase.MatReadError: Mat file appears to be empty

Does anybody meet the same problem?  Know the reason or know how to solve this? ?  It is important and urgent for me!!! Thanks a lot!!!

**Pull request merged**